lib/integrations/awsoidc/listec2_test.go

```diff
@@ -0,0 +1,233 @@
+		})
+		require.NoError(t, err)
+		require.NotEmpty(t, resp.NextToken)
+		require.Len(t, resp.Servers, pageSize)
+		nextPageToken = resp.NextToken
+
+		// Third page must return only the remaining Servers and an empty nextToken
+		resp, err = ListEC2(ctx, mockListClient, ListEC2Request{
+			Region:      "us-east-1",
+			Integration: "myintegration",
+			NextToken:   nextPageToken,
+		})
+		require.NoError(t, err)
+		require.Empty(t, resp.NextToken)
+		require.Len(t, resp.Servers, 3)
+	})
+
+	for _, tt := range []struct {
+		name          string
+		req           ListEC2Request
+		mockInstances []ec2Types.Instance
+		errCheck      func(error) bool
+		respCheck     func(*testing.T, *ListEC2Response)
+	}{
+		{
+			name: "valid for listing instances",
+			req: ListEC2Request{
+				Region:      "us-east-1",
+				Integration: "myintegration",
+				NextToken:   "",
+			},
+			mockInstances: []ec2Types.Instance{{
+				PrivateDnsName:   aws.String("my-private-dns.compute.aws"),
+				InstanceId:       aws.String("i-123456789abcedf"),
+				VpcId:            aws.String("vpc-abcd"),
+				PrivateIpAddress: aws.String("172.31.1.1"),
+			},
+			},
+			respCheck: func(t *testing.T, ldr *ListEC2Response) {
+				require.Len(t, ldr.Servers, 1, "expected 1 server, got %d", len(ldr.Servers))
+				require.Empty(t, ldr.NextToken, "expected an empty NextToken")
+
+				expectedServer := &types.ServerV2{
+					Kind:    "node",
+					Version: "v2",
+					SubKind: "openssh-ec2-ice",
+					Metadata: types.Metadata{
+						Labels: map[string]string{
+							"account-id": "123456789012",
+							"region":     "us-east-1",
+						},
+						Namespace: "default",
+					},
+					Spec: types.ServerSpecV2{
+						Addr:     "172.31.1.1:22",
+						Hostname: "my-private-dns.compute.aws",
+						CloudMetadata: &types.CloudMetadata{
+							AWS: &types.AWSInfo{
+								AccountID:   "123456789012",
+								InstanceID:  "i-123456789abcedf",
+								Region:      "us-east-1",
+								VPCID:       "vpc-abcd",
+								Integration: "myintegration",
+							},
+						},
+					},
+				}
+				require.Empty(t, cmp.Diff(expectedServer, ldr.Servers[0], cmpopts.IgnoreFields(types.ServerV2{}, "Metadata.Name")))
+			},
+			errCheck: noErrorFunc,
+		},
+		{
+			name: "no region",
+			req: ListEC2Request{
+				Integration: "myintegration",
+			},
+			errCheck: func(err error) bool {
+				return trace.IsBadParameter(err)
+			},
+		},
+		{
+			name: "no integration",
+			req: ListEC2Request{
+				Region: "us-east-1",
+			},
+			errCheck: func(err error) bool {
+				return trace.IsBadParameter(err)
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			mockListClient := &mockListEC2Client{
+				pageSize:     pageSize,
+				accountID:    "123456789012",
+				ec2Instances: tt.mockInstances,
+			}
+			resp, err := ListEC2(ctx, mockListClient, tt.req)
+			require.True(t, tt.errCheck(err), "unexpected err: %v", err)
+			if err != nil {
+				return
+			}
+
+			tt.respCheck(t, resp)
```

Comment:
nit: Probably opinion, but I find a few times `respCheck` like function to not being called because of a return like that. IMO this is less error-prone:
```suggestion
			if tt.respCheck != nil {
				tt.respCheck(t, resp)
			}
```

Reply:
The assumption is that when there's an error, it will not call the `respCheck`
That is not very explicit when creating the table test.

I think it usually works by convention, but I'll replace the check to be explicit :+1: